### ISSUE TITLE
Should not be using conda run to activate default conda envs

### ISSUE THREAD

**Issue by grallewellyn** [MAAP-Project/ZenHub#967]
"Generally speaking though, I don't think we should be taking this approach. The point of using conda run is not for it to be used as a means of activating a "default" conda env in the container. Users should be using conda run themselves, in place of conda activate or source activate."
- Chuck Daniels
More info on this ticket: https://github.com/MAAP-Project/Community/issues/965
cc @anilnatha 

Just need to change conda run here in jupyterlab3/dockerfile [here](https://github.com/MAAP-Project/maap-workspaces/blob/main/jupyterlab3/docker/Dockerfile#L10-L20): 
```
FROM jupyterlab_base as workspace_base_pangeo
SHELL ["conda", "run", "-n", "pangeo", "/bin/bash", "-c"]

FROM jupyterlab_base as workspace_base_vanilla
SHELL ["conda", "run", "-n", "vanilla", "/bin/bash", "-c"]

FROM jupyterlab_base as workspace_base_r
SHELL ["conda", "run", "-n", "r", "/bin/bash", "-c"]

FROM jupyterlab_base as workspace_base_isce3
SHELL ["conda", "run", "-n", "isce3", "/bin/bash", "-c"]
```
Because in the [base_images/ dockerfiles](https://github.com/MAAP-Project/maap-workspaces/blob/main/base_images/isce3/docker/Dockerfile#L22) we don't need this line:
`SHELL ["conda", "run", "-n", "isce3", "/bin/bash", "-c"]` once we package maap-py in this next release

**Comment by grallewellyn** [comment]
Sujen suggests we use conda init instead of conda run 
This means when we do a pip install, we need to specify the absolute path to pip, ie opt/conda/...
But soon we shouldn't be doing any pip installs in the dockerfiles

**Comment by chuckwondo** [comment]
Agreed. We should definitely add `conda init` to the `Dockerfile`. If that wasn't already in place, then this very likely explains the issue about not being able to run `conda activate`. Running `conda init` will populate `~/.bash_profile` as required in order to allow a user to run `conda activate`.

**Comment by chuckwondo** [comment]
Additionally, after running `conda init` to populate `~/.bash_profile`, if you also want to activate a "default" environment for the user, you should then append an activation command to `~/.bash_profile`, probably like so:

```plain
cat >> ~/.bash_profile <<EOF
conda activate vanilla
EOF
```

Of course, depending on the image, replace `vanilla` with the appropriate conda environment name.

Note that it is important that the command above is executed _after_ `conda init` so that the "activate" command appears _after_ the code inserted by the `conda init` command.

**Comment by grallewellyn** [comment]
In the base_images dockerfile, I am running `RUN conda init bash` then adding "conda activate vanilla" to the .bash_profile 

`RUN source activate vanilla` or `SHELL ["source", "activate", "vanilla"]` doesn’t cause following conda installs to go into the vanilla conda environment. Did you mean we should activate vanilla within our dockerfiles, or by default in the ADE? 
We could just specify the environment name in our pip/ conda installs like `RUN conda install --name vanilla -y scipy` and `RUN /opt/conda/envs/vanilla/bin/pip install scipy` which install into vanilla properly

**Comment by chuckwondo** [comment]
What I mean is that you should put this in the `Dockerfile`, replacing `vanilla` with whatever the appropriate conda environment name is:

```plain
RUN conda init && cat >> ~/.bashrc <<EOF
conda activate vanilla
EOF
```

The `conda init` command writes to `~/.bashrc`, so we simply append `conda activate vanilla` _after_ the changes that `conda init` makes. Therefore, the indicated environment is activated for the user by default, but the user is also free to use `conda activate` to activate other environments, if they wish, and `conda activate` will work because `conda init` was the previously missing piece.

**Comment by chuckwondo** [comment]
This also means that you do _not_ need a `SHELL` entry like you mentioned earlier:

```plain
SHELL ["conda", "run", "-n", "vanilla", "/bin/bash", "-c"]
```

**Comment by grallewellyn** [comment]
We are currently activating the vanilla, etc. conda environments by default for the user [here](https://github.com/MAAP-Project/maap-workspaces/blob/main/jupyterlab3/entrypoint.sh#L129-L130):
```
# Set up default conda environment for terminals
echo ". /opt/conda/etc/profile.d/conda.sh ; conda activate $WORKSPACE_TYPE" > /etc/profile.d/init_conda.sh
```
Are you suggesting we change from doing it this way? Why? Would we rather be editing `.bashrc` instead of `init_conda.sh`?

To have our conda/ pip installs go into the correct conda environment when we are creating the images should we `RUN conda install --name vanilla -y scipy` and `RUN /opt/conda/envs/vanilla/bin/pip install scipy` if we aren't going to run `SHELL ["conda", "run", "-n", "vanilla", "/bin/bash", "-c"]` anymore? Just modifying the `.bashrc` in the dockerfile is not enough to have installs in the Dockerfile go into the correct conda environment

**Comment by grallewellyn** [comment]
Nevermind on this, looks like doing conda/ pip installs in our dockerfiles will no longer be relevant with the coming release because we are planning to extract maap-py and jupyter extension installs into the environment.yml
I will go through the jupyterlab3/ dockerfile and check what commands were relying on the vanilla, etc. conda environment being activated

**Comment by grallewellyn** [comment]
@chuckwondo For the maap-workspaces jupyterlab3/dockerfile, adding `ENV PATH="$PATH:/opt/conda/envs/$WORKSPACE_TYPE/bin"` to the dockerfile allows all following `npm` and `jupyter` commands to correctly install into the correct conda environment (or just correctly execute e.g. in the case of disabling Jupyter announcements). This approach allows us to remove all instances of `conda run`. 

Would this be an okay approach to avoid using `conda run`?

We are also exporting the path in our entrypoint.sh with `export PATH=$PATH:/opt/conda/envs/$WORKSPACE_TYPE/bin` and I believe we will need both lines, I can test more if we decide to go with this approach

**Comment by chuckwondo** [comment]
> @chuckwondo For the maap-workspaces jupyterlab3/dockerfile, adding `ENV PATH="$PATH:/opt/conda/envs/$WORKSPACE_TYPE/bin"` to the dockerfile allows all following `npm` and `jupyter` commands to correctly install into the correct conda environment (or just correctly execute e.g. in the case of disabling Jupyter announcements). This approach allows us to remove all instances of `conda run`.
> 
> Would this be an okay approach to avoid using `conda run`?
> 
> We are also exporting the path in our entrypoint.sh with `export PATH=$PATH:/opt/conda/envs/$WORKSPACE_TYPE/bin` and I believe we will need both lines, I can test more if we decide to go with this approach

Yes, I said earlier that you should _definitely not_ use `conda run`.

Also, you should not need to export `PATH`.

Again, here's what I believe you should put into the `Dockerfile` (from an earlier comment):

> What I mean is that you should put this in the `Dockerfile`, replacing `vanilla` with whatever the appropriate conda environment name is:
> 
> ```
> RUN conda init && cat >> ~/.bashrc <<EOF
> conda activate vanilla
> EOF
> ```
> 
> The `conda init` command writes to `~/.bashrc`, so we simply append `conda activate vanilla` _after_ the changes that `conda init` makes. Therefore, the indicated environment is activated for the user by default, but the user is also free to use `conda activate` to activate other environments, if they wish, and `conda activate` will work because `conda init` was the previously missing piece.

**Comment by grallewellyn** [comment]
I did put conda init in the dockerfile, but just running 
```
RUN conda init && cat >> ~/.bashrc <<EOF
conda activate vanilla
EOF
```
is not enough to have following Docker commands install into the correct conda enviroment. We have to either export the PATH or activate the conda environment. We don't have any pip/ conda installs in the Dockerfiles, but we do have jupyter and npm commands. In our MAAP meeting, we discussed we would rather go the route of activating the conda environment with source activate, but /bin/sh doesn’t recognize source, so I was able to get it to run as
```
RUN /bin/bash -c 'source activate vanilla && \
    npm install typescript -g && \
    npm install jupyterlab-plotly@5.5.0 && \
    jupyter labextension disable @jupyterlab/apputils-extension:announcements && \
    jupyter server extension enable jupyter_server_extension'
```

**Comment by chuckwondo** [comment]
There's no need to put `/bin/bash -c` at the front of your `RUN` commands. That's what setting `SHELL` is for.

Here's the type of thing you need to do:

```Dockerfile
...

SHELL ["/bin/bash", "-c"]
WORKDIR /projects
RUN sed -i -e 's/\/root/\/projects/g' /etc/passwd

...

RUN conda init && cat >> ~/.bashrc <<EOF
conda activate vanilla
EOF

...

RUN conda run -n vanilla npm install typescript -g \
    && conda run -n vanilla npm install jupyterlab-plotly@5.5.0 \
    && conda run -n vanilla jupyter labextension disable @jupyterlab/apputils-extension:announcements \
    && conda run -n vanilla jupyter server extension enable jupyter_server_extension
```

Note that the `conda activate vanilla` line added to `~/.bashrc` is _NOT_ to help with anything else going on in the Dockerfile itself. That line is strictly for the _user_, such that when they open a terminal, the "vanilla" environment will automatically be activated for them.

However, this does not prevent the user from also creating and `conda activate`ing their own environments in addition to the default "vanilla" environment. The `conda init` command is what allows the user to use `conda activate` correctly.

**Comment by grallewellyn** [comment]
If we would rather set the`SHELL` to execute this, then we can do that. You mentioned before we shouldn't be using `bin/bash` with `conda run`, conversation [here](https://github.com/MAAP-Project/Community/issues/965), so I didn't know if we wanted to set the `SHELL` so all following dockerlines were ran with `/bin/bash` 

Why are you suggesting we do
```
RUN conda run -n vanilla npm install typescript -g \
    && conda run -n vanilla npm install jupyterlab-plotly@5.5.0 \
    && conda run -n vanilla jupyter labextension disable @jupyterlab/apputils-extension:announcements \
    && conda run -n vanilla jupyter server extension enable jupyter_server_extension
```
instead of 
```
SHELL ["/bin/bash", "-c"]
RUN source activate vanilla && \
    npm install typescript -g && \
    npm install jupyterlab-plotly@5.5.0 && \
    jupyter labextension disable @jupyterlab/apputils-extension:announcements && \
    jupyter server extension enable jupyter_server_extension
```
The MAAP team has discussed that we would like to remove instances of `conda run` from the dockerfile

My dockerfile won't run with 
```
RUN conda init && cat >> ~/.bashrc <<EOF
conda activate vanilla
EOF
```
because `unknown instruction: CONDA`, so I am running `RUN conda init && echo "conda activate vanilla" >> ~/.bashrc` instead

**Comment by chuckwondo** [comment]
> If we would rather set the`SHELL` to execute this, then we can do that. You mentioned before we shouldn't be using `bin/bash` with `conda run`, conversation [here](https://github.com/MAAP-Project/Community/issues/965), so I didn't know if we wanted to set the `SHELL` so all following dockerlines were ran with `/bin/bash`

Using `/bin/bash` with `conda run` is perfectly fine as shown below, where `conda run` is used within a `RUN` directive. The problem is using `"conda", "run"` within a `SHELL` directive. I was saying that you do _not_ want to do something like this: `SHELL ["conda", "run", "-n", "vanilla", "/bin/bash", "-c"]`

> 
> Why are you suggesting we do
> 
> ```
> RUN conda run -n vanilla npm install typescript -g \
>     && conda run -n vanilla npm install jupyterlab-plotly@5.5.0 \
>     && conda run -n vanilla jupyter labextension disable @jupyterlab/apputils-extension:announcements \
>     && conda run -n vanilla jupyter server extension enable jupyter_server_extension
> ```
> 
> instead of
> 
> ```
> SHELL ["/bin/bash", "-c"]
> RUN source activate vanilla && \
>     npm install typescript -g && \
>     npm install jupyterlab-plotly@5.5.0 && \
>     jupyter labextension disable @jupyterlab/apputils-extension:announcements && \
>     jupyter server extension enable jupyter_server_extension
> ```
> 
> The MAAP team has discussed that we would like to remove instances of `conda run` from the dockerfile

Using `conda run` in a `RUN` command is perfectly fine, but it requires running `conda init` beforehand (which is why I included that separately in my previous comment). The problem I was referring to is that you do not want to use `"conda", "run"` within a `SHELL` directive.

However, if using `source activate vanilla` (instead of `conda run`) works for you in your `RUN` command, that's fine too.

> 
> My dockerfile won't run with
> 
> ```
> RUN conda init && cat >> ~/.bashrc <<EOF
> conda activate vanilla
> EOF
> ```
> 
> because `unknown instruction: CONDA`, so I am running `RUN conda init && echo "conda activate vanilla" >> ~/.bashrc` instead

I suspect you unknowingly copy-pasted an invisible control character. What I gave you works perfectly fine for me, but if using `echo` instead worked for you, that's fine too.
